### PR TITLE
Enforce determinism in ArDoCo by architecture tests

### PR DESCRIPTION
**Pull request by dfuchss**
This PR introduces mechanisms to enforce determinism in ArDoCo.
Therefore, the use of HashSets, Sets, HashMaps, and Maps has been restricted and requires OptIn for classes.

### CODE REVIEW

File: stages/text-extraction/src/main/java/edu/kit/kastel/mcse/ardoco/core/textextraction/OriginalTextStateStrategy.java [hunk 1]
```diff
@@ -2,43 +2,35 @@
...
     OriginalTextStateStrategy(TextStateImpl textState) {
         super.setTextState(textState);
     }
 
     @Override
     public NounMapping addOrExtendNounMapping(Word word, MappingKind kind, Claimant claimant, double probability, ImmutableList<String> surfaceForms) {
 
-        NounMapping disposableNounMapping = new NounMappingImpl(System.currentTimeMillis(), Sets.immutable.with(word), kind, claimant, probability,
-                Lists.immutable.with(word), surfaceForms);
+        NounMapping disposableNounMapping = new NounMappingImpl(NounMappingImpl.getNextCreationTime(), SortedSets.immutable.with(word), kind, claimant,
```
Comment (Gram21) [inline stages/text-extraction/src/main/java/edu/kit/kastel/mcse/ardoco/core/textextraction/OriginalTextStateStrategy.java:32]: Isn't it better to offer a second constructor (or change the existing one) that has the first parameter (creationTime) removed? It is odd when a class calls the constructor and needs to access a date from the class it tries to construct.

File: stages/text-extraction/src/main/java/edu/kit/kastel/mcse/ardoco/core/textextraction/informants/MappingCombinerInformant.java [hunk 2]
```diff
@@ -41,6 +42,9 @@ private void combineSimilarPhraseMappings(TextState textState) {
         for (PhraseMapping phraseMapping : phraseMappings) {
             ImmutableList<PhraseMapping> similarPhraseMappings = phraseMappings.select(p -> SimilarityUtils.getPhraseMappingSimilarity(textState, phraseMapping,
                     p, PhraseMappingAggregatorStrategy.MAX_SIMILARITY) > minCosineSimilarity);
+            // Remove the phrase mapping from the list of similar phrase mappings
+            // TODO This breaks the logic but seems to be logical ..
```
Comment (Gram21) [inline stages/text-extraction/src/main/java/edu/kit/kastel/mcse/ardoco/core/textextraction/informants/MappingCombinerInformant.java:46]: Is this TODO and the commented code up to date? Can it be removed?

File: tests/tests-tlr/src/test/java/edu/kit/kastel/mcse/ardoco/core/tests/integration/TraceabilityLinkRecoveryEvaluation.java [hunk 3]
```diff
@@ -159,8 +158,8 @@ protected EvaluationResults<String> calculateEvaluationResults(ArDoCoResult arDo
         ImmutableList<String> results = createTraceLinkStringList(arDoCoResult);
         Assertions.assertFalse(results.isEmpty());
 
-        Set<String> distinctTraceLinks = new HashSet<>(results.castToCollection());
-        Set<String> distinctGoldStandard = new HashSet<>(goldStandard.castToCollection());
+        Set<String> distinctTraceLinks = new java.util.LinkedHashSet<>(results.castToCollection());
+        Set<String> distinctGoldStandard = new java.util.LinkedHashSet<>(goldStandard.castToCollection());
```
Comment (Gram21) [inline tests/tests-tlr/src/test/java/edu/kit/kastel/mcse/ardoco/core/tests/integration/TraceabilityLinkRecoveryEvaluation.java:162]: `import java.util.LinkedHashSet`?

**Review by Gram21 — CHANGES_REQUESTED** [review]

File: stages/text-extraction/src/main/java/edu/kit/kastel/mcse/ardoco/core/textextraction/informants/MappingCombinerInformant.java (hunk 2 above)
Comment (dfuchss) [reply stages/text-extraction/src/main/java/edu/kit/kastel/mcse/ardoco/core/textextraction/informants/MappingCombinerInformant.java:46]: It's a finding of mine. I'd not remove that comment.

File: framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/api/connectiongenerator/ConnectionState.java [hunk 4]
```diff
@@ -64,7 +66,7 @@ public interface ConnectionState extends IConfigurable {
      * @return list of tracelinks within this connection state
      */
     default ImmutableSet<SadSamTraceLink> getTraceLinks() {
```
[the remark this answers is not in the document]
Comment (dfuchss) [reply framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/api/connectiongenerator/ConnectionState.java:68]: Does not change anything . Reverted.

**Review by Gram21 — APPROVED** [review]

**Pull request merged**